[v2 stability issue] GraphQL Error Unknown field `modified` on type `NPMPackage` when running gatsby's `www` locally

## Description

running the exact same code yesterday as I did today, and it's unchanged, but now I am getting this error:

```bash
npm run develop

> gatsby-starter-default@1.1.0-alpha.03b9df85 develop /Users/swyx/gatsby/StarterShowcase/gatsby/www
> gatsby develop

success delete html files from previous builds — 0.244 s
success open and validate gatsby-config — 0.004 s
success copy gatsby files — 0.034 s
success onPreBootstrap — 1.460 s
⢀ source and transform nodesThe "deleteNodes" action is now deprecated and will be removed in Gatsby v3. Please use "deleteNode" instead.
The "deleteNodes" action is now deprecated and will be removed in Gatsby v3. Please use "deleteNode" instead.
success source and transform nodes — 3.817 s
success building schema — 1.147 s
success createPages — 3.678 s
success createPagesStatefully — 0.241 s
success onPreExtractQueries — 0.008 s
success update schema — 2.190 s
warning There are conflicting field types in your data. GraphQL schema will omit those fields.
NPMPackage.created:
NPMPackage.modified:
MarkdownRemark.fields.starterShowcase.githubData.repository:
SitePluginConnection.packageJson.author:
NPMPackageConnection.created:
NPMPackageConnection.modified:
MarkdownRemarkConnection.fields.starterShowcase.githubData.repository:
GraphQL Error Unknown field `modified` on type `NPMPackage`

  file: /Users/swyx/gatsby/StarterShowcase/gatsby/www/src/templates/template-docs-local-packages.js

   9 |       }
  10 |       ...MarkdownPageFooter
  11 |     }
  12 |     npmPackage(slug: { eq: $slug }) {
  13 |       name
  14 |       keywords
  15 |       lastPublisher {
  16 |         name
  17 |         avatar
  18 |       }
> 19 |       modified
     |       ^
  20 |     }
  21 |   }
  22 |

success extract queries from components — 0.289 s
⠁ (node:73420) DeprecationWarning: Passing lineNumber and colNumber is deprecated to @babel/code-frame. Please use `codeFrameColumns`.
success run graphql queries — 0.399 s
success write out page data — 0.023 s
success write out redirect data — 0.001 s
success onPostBootstrap — 0.001 s

info bootstrap finished - 18.178 s

ℹ ｢wdm｣: wait until bundle finished: /starters/test
 ERROR  Failed to compile with 2 errors                                                                                                                    09:39:57

 error  in ./src/templates/template-docs-local-packages.js

Module Error (from ./node_modules/eslint-loader/index.js):

/Users/swyx/gatsby/StarterShowcase/gatsby/www/src/templates/template-docs-local-packages.js
  96:7  error  Cannot query field "modified" on type "NPMPackage"  graphql/template-strings

✖ 1 problem (1 error, 0 warnings)


 @ ./.cache/sync-requires.js 17:77-179
 @ ./.cache/root.js
 @ ./.cache/app.js
 @ multi ./node_modules/react-hot-loader/patch.js (webpack)-hot-middleware/client.js?path=http://localhost:8000/__webpack_hmr&reload=true&overlay=false ./.cache/app
```

to be clear this is unrelated code to [starter showcase](https://github.com/gatsbyjs/gatsby/pull/5831) so this could very well happen to anyone else running `www` on the v2 branch, potentially anyone using v2. just logging this error here for now so that others can find it, i have no idea what the cause is yet.

**NOTE: when i kill the process, and simply `npm run develop` again, the code works and the site builds with no errors. this is very definitely a stability issue of some sort.**

### Steps to reproduce

```
git checkout v2
cd www
npm run develop
```

### Expected result

website should run

### Actual result

doesn't run

### Environment


  System:
    OS: macOS High Sierra 10.13.3
    CPU: x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 9.4.0 - /usr/local/bin/node
    Yarn: 1.6.0 - /usr/local/bin/yarn
    npm: 6.0.1 - /usr/local/bin/npm
  Browsers:
    Chrome: 66.0.3359.181
    Firefox: 59.0.2
    Safari: 11.0.3
  npmPackages:
    gatsby: next => 2.0.0-alpha.51 
    gatsby-image: next => 2.0.0-alpha.5 
    gatsby-plugin-canonical-urls: next => 1.0.14-10 
    gatsby-plugin-catch-links: next => 2.0.1-11 
    gatsby-plugin-feed: next => 1.3.19-11 
    gatsby-plugin-fullstory: ^1.0.4-5 => 1.0.4-7 
    gatsby-plugin-glamor: next => 2.0.0-alpha.13 
    gatsby-plugin-google-analytics: next => 1.0.20-10 
    gatsby-plugin-lodash: next => 3.0.0-alpha.12 
    gatsby-plugin-mailchimp: ^2.2.3 => 2.2.3 
    gatsby-plugin-manifest: next => 2.0.1-17 
    gatsby-plugin-netlify: next => 1.0.19-11 
    gatsby-plugin-nprogress: next => 2.0.0-alpha.12 
    gatsby-plugin-offline: next => 2.0.0-alpha.15 
    gatsby-plugin-react-helmet: next => 2.0.12-7 
    gatsby-plugin-sharp: next => 2.0.0-alpha.19 
    gatsby-plugin-sitemap: next => 1.2.14-10 
    gatsby-plugin-twitter: next => 1.0.18-11 
    gatsby-plugin-typography: next => 2.0.1-11 
    gatsby-remark-autolink-headers: next => 1.4.13-11 
    gatsby-remark-copy-linked-files: next => 1.5.28-10 
    gatsby-remark-images: next => 2.0.0-alpha.21 
    gatsby-remark-prismjs: next => 1.2.16-11 
    gatsby-remark-responsive-iframe: next => 1.4.18-10 
    gatsby-remark-smartypants: next => 1.4.12-11 
    gatsby-source-filesystem: next => 2.0.0-alpha.15 
    gatsby-source-npm-package-search: ^1.0.8 => 1.0.8 
    gatsby-transformer-csv: next => 1.3.9-10 
    gatsby-transformer-documentationjs: next => 1.4.10-10 
    gatsby-transformer-remark: next => 2.0.1-13 
    gatsby-transformer-sharp: next => 2.0.1-13 
    gatsby-transformer-yaml: next => 2.0.1-11 
  npmGlobalPackages:
    gatsby-cli: 1.1.58
    gatsby: 1.9.231

### File contents (if changed)

n/a

--- a/packages/gatsby/src/schema/__tests__/data-tree-utils-test.js
+++ b/packages/gatsby/src/schema/__tests__/data-tree-utils-test.js
@@ -205,6 +205,154 @@ describe(`Gatsby data tree utils`, () => {
     expect(example.numbers.length).toBe(1)
     expect(example.numbers[0]).toBe(2.5)
   })
+
+  it(`handles mix of date strings and date objects`, () => {
+    let example
+
+    // should be valid
+    example = getExampleValues({
+      nodes: [
+        { date: new Date(`2017-12-01T14:59:45.600Z`) },
+        { date: `2017-01-12T18:13:38.326Z` },
+      ],
+    })
+    expect(example.date).not.toBe(INVALID_VALUE)
+
+    // should be invalid (string is not a date)
+    example = getExampleValues({
+      nodes: [
+        { date: new Date(`2017-12-01T14:59:45.600Z`) },
+        { date: `This is not a date!!!!!!` },
+      ],
+    })
+    expect(example.date).toBe(INVALID_VALUE)
+
+    // should be valid - reversed order
+    example = getExampleValues({
+      nodes: [
+        { date: `2017-01-12T18:13:38.326Z` },
+        { date: new Date(`2017-12-01T14:59:45.600Z`) },
+      ],
+    })
+    expect(example.date).not.toBe(INVALID_VALUE)
+
+    // should be invalid (string is not a date) - reversed order
+    example = getExampleValues({
+      nodes: [
+        { date: `This is not a date!!!!!!` },
+        { date: new Date(`2017-12-01T14:59:45.600Z`) },
+      ],
+    })
+    expect(example.date).toBe(INVALID_VALUE)
+  })
+
+  it(`handles arrays with mix of date strings and date objects`, () => {
+    let example
+
+    // should be valid - separate arrays of unique types
+    example = getExampleValues({
+      nodes: [
+        { dates: [new Date(`2017-12-01T14:59:45.600Z`)] },
+        { dates: [`2017-01-12T18:13:38.326Z`] },
+      ],
+    })
+    expect(example.dates).not.toBe(INVALID_VALUE)
+
+    // should be invalid - separate arrays of unique types (string is not a date)
+    example = getExampleValues({
+      nodes: [
+        { dates: [new Date(`2017-12-01T14:59:45.600Z`)] },
+        { dates: [`This is not a date!!!!!!`] },
+      ],
+    })
+    expect(example.dates).toBe(INVALID_VALUE)
+
+    // should be valid - single array of mixed types
+    example = getExampleValues({
+      nodes: [
+        {
+          dates: [
+            new Date(`2017-12-01T14:59:45.600Z`),
+            `2017-01-12T18:13:38.326Z`,
+          ],
+        },
+      ],
+    })
+    expect(example.dates).not.toBe(INVALID_VALUE)
+
+    // should be invalid - single array of mixed types (string is not a date)
+    example = getExampleValues({
+      nodes: [
+        {
+          dates: [
+            new Date(`2017-12-01T14:59:45.600Z`),
+            `This is not a date!!!!!!`,
+          ],
+        },
+      ],
+    })
+    expect(example.dates).toBe(INVALID_VALUE)
+
+    // should be valid - separate arrays of both unique types and mixed types
+    example = getExampleValues({
+      nodes: [
+        {
+          dates: [
+            new Date(`2017-12-01T14:59:45.600Z`),
+            `2017-01-12T18:13:38.326Z`,
+          ],
+        },
+        { dates: [new Date(`2017-12-01T14:59:45.600Z`)] },
+        { dates: [`2017-01-12T18:13:38.326Z`] },
+      ],
+    })
+    expect(example.dates).not.toBe(INVALID_VALUE)
+
+    // should be valid - separate arrays of both unique types and mixed types (string is not a date) #1
+    example = getExampleValues({
+      nodes: [
+        {
+          dates: [
+            new Date(`2017-12-01T14:59:45.600Z`),
+            `This is not a date!!!!!!`,
+          ],
+        },
+        { dates: [new Date(`2017-12-01T14:59:45.600Z`)] },
+        { dates: [`2017-01-12T18:13:38.326Z`] },
+      ],
+    })
+    expect(example.dates).toBe(INVALID_VALUE)
+
+    // should be valid - separate arrays of both unique types and mixed types (string is not a date) #2
+    example = getExampleValues({
+      nodes: [
+        {
+          dates: [
+            new Date(`2017-12-01T14:59:45.600Z`),
+            `2017-01-12T18:13:38.326Z`,
+          ],
+        },
+        { dates: [new Date(`2017-12-01T14:59:45.600Z`)] },
+        { dates: [`This is not a date!!!!!!`] },
+      ],
+    })
+    expect(example.dates).toBe(INVALID_VALUE)
+
+    // should be valid - separate arrays of both unique types and mixed types (string is not a date) #2
+    example = getExampleValues({
+      nodes: [
+        {
+          dates: [
+            new Date(`2017-12-01T14:59:45.600Z`),
+            `This is not a date!!!!!!`,
+          ],
+        },
+        { dates: [new Date(`2017-12-01T14:59:45.600Z`)] },
+        { dates: [`This is not a date!!!!!!`] },
+      ],
+    })
+    expect(example.dates).toBe(INVALID_VALUE)
+  })
 })
 
 describe(`Type conflicts`, () => {

--- a/packages/gatsby/src/schema/data-tree-utils.js
+++ b/packages/gatsby/src/schema/data-tree-utils.js
@@ -6,6 +6,7 @@ const invariant = require(`invariant`)
 
 const createKey = require(`./create-key`)
 const { typeConflictReporter } = require(`./type-conflict-reporter`)
+const DateType = require(`./types/type-date`)
 
 import type { TypeEntry } from "./type-conflict-reporter"
 
@@ -56,6 +57,62 @@ const extractTypes = value => {
   }
 }
 
+const isMixOfDateObjectsAndDateStrings = (
+  values: any[],
+  uniqueTypes: string[]
+): boolean => {
+  if (
+    uniqueTypes.length === 2 &&
+    uniqueTypes.includes(`string`) &&
+    uniqueTypes.includes(`date`)
+  ) {
+    const allValuesAreDates = values.every(value => {
+      if (typeOf(value) === `date`) return true
+
+      // use infer checker to determine if string is a date
+      return DateType.shouldInfer(value)
+    })
+
+    return allValuesAreDates
+  }
+
+  return false
+}
+
+const conflictIsValidSpecialCase = (
+  entries: TypeEntry[],
+  entriesOfUniqueType: TypeEntry[]
+): boolean => {
+  const isConsistentlyScalarOrArray = entriesOfUniqueType.every(
+    entry =>
+      entry.arrayTypes.length > 0 ===
+      entriesOfUniqueType[0].arrayTypes.length > 0
+  )
+  if (isConsistentlyScalarOrArray) {
+    // Get values and run them through special cases, to see if there actually
+    // is a conflict. This is done so late in process, because those checks
+    // would be expensive to do earlier during extraction, so we do them
+    // only when we have potential conflict.
+
+    let uniqueTypes: string[]
+    let values: any[]
+    const isArray = entriesOfUniqueType[0].arrayTypes.length > 0
+
+    if (isArray) {
+      values = _.flatten(entries.map(entry => entry.value))
+      uniqueTypes = _.uniq(
+        _.flatten(entriesOfUniqueType.map(entry => entry.arrayTypes))
+      )
+    } else {
+      values = entries.map(entry => entry.value)
+      uniqueTypes = entriesOfUniqueType.map(entry => entry.type)
+    }
+
+    return isMixOfDateObjectsAndDateStrings(values, uniqueTypes)
+  }
+  return false
+}
+
 const getExampleScalarFromArray = values =>
   _.reduce(
     values,
@@ -87,10 +144,13 @@ const extractFromEntries = (
     entriesOfUniqueType[0].arrayTypes.length > 1
   ) {
     // there is multiple types or array of multiple types
-    if (selector) {
-      typeConflictReporter.addConflict(selector, entriesOfUniqueType)
+    // that aren't handled by any special case
+    if (!conflictIsValidSpecialCase(entries, entriesOfUniqueType)) {
+      if (selector) {
+        typeConflictReporter.addConflict(selector, entriesOfUniqueType)
+      }
+      return INVALID_VALUE
     }
-    return INVALID_VALUE
   }
 
   // Now we have entries of single type, we can merge them

